onadata/libs/utils/xml.py

```diff
@@ -0,0 +1,183 @@
+from __future__ import annotations
+
+import re
+from typing import Optional, Union
+from xml.dom import Node
+
+from defusedxml import minidom
+from django.db.models import F, Q
+from django.db.models.query import QuerySet
+
+from onadata.apps.form_disclaimer.models import FormDisclaimer
+
+
+class XMLFormWithDisclaimer:
+
+    # TODO merge this with KPI when Kobocat becomes a Django-app
+    def __init__(self, obj: Union['logger.XForm']):
+        self._object = obj
+        self._unique_id = obj.id_string
+        self._add_disclaimer()
+
+    def get_object(self):
+        return self._object
+
+    def _add_disclaimer(self):
+
+        if not (disclaimers := self._get_disclaimers(self._object)):
+            return
+
+        if not (value := self._get_translations(disclaimers)):
+            return
+
+        translated, disclaimers_dict, default_language_code = value
+
+        self._root_node = minidom.parseString(self._object.xml)
+
+        if translated:
+            self._add_translation_nodes(disclaimers_dict, default_language_code)
+
+        self._add_instance_and_bind_nodes()
+        self._add_disclaimer_input(
+            translated, disclaimers_dict, default_language_code
+        )
+
+        self._object.xml = self._root_node.toxml(encoding='utf-8').decode()
+
+    def _add_instance_and_bind_nodes(self):
+        # Search for main <model> node
+        model_node = [
+            n
+            for n in self._root_node.getElementsByTagName('h:head')[0].childNodes
+            if n.nodeType == Node.ELEMENT_NODE and n.tagName == 'model'
+        ][0]
+
+        # Inject <bind nodeset /> inside <model odk:xforms-version="1.0.0">
+        bind_node = self._root_node.createElement('bind')
+        bind_node.setAttribute(
+            'nodeset', f'/{self._unique_id}/_{self._unique_id}__disclaimer'
+        )
+        bind_node.setAttribute('readonly', 'true()')
+        bind_node.setAttribute('required', 'false()')
+        bind_node.setAttribute('type', 'string')
+        bind_node.setAttribute('relevant', 'false()')
+        model_node.appendChild(bind_node)
+
+        # Inject note node inside <{self._unique_id}>
+        instance_node = model_node.getElementsByTagName('instance')[0]
+        instance_node = instance_node.getElementsByTagName(self._unique_id)[0]
```

Comment:
assuming this is okay because iirc kobocat never respected https://xlsform.org/en/#specify-xforms-root-node-name